rewrite-core/src/main/java/org/openrewrite/RecipeScheduler.java

```diff
@@ -308,10 +308,42 @@ default <S extends SourceFile> List<S> scheduleVisit(RecipeRunStats runStats,
         // of a type that is in the original set of source files (e.g. only XML files are given, and the
         // recipe generates Java code).
         List<SourceFile> afterWidened;
+        final List<Boolean> lastSingleSourceApplicableTestResult = singleSourceApplicableTestResult;
+        final List<Boolean> newSingleSourceApplicableTestResult = lastSingleSourceApplicableTestResult;
         try {
             long ownVisitStartTime = System.nanoTime();
+
+            // If any single source in the set matches single source applicability test, we need to run all files
+            // together, but just keep the results generated from the files which passed the single source applicability test.
+            if (hasSingleSourceApplicableTest) {
+                boolean anyFilePassedSingleApplicableTest = false;
+                for (Boolean b : singleSourceApplicableTestResult) {
+                    if (b) {
+                        anyFilePassedSingleApplicableTest = true;
+                        break;
+                    }
+                }
+                if (!anyFilePassedSingleApplicableTest) {
+                    return after;
+                }
+            }
+
             //noinspection unchecked
             afterWidened = recipe.visit((List<SourceFile>) after, ctx);
+            final List<SourceFile> afterWidenedRef = afterWidened;
+
+            // update/re-mapping single source applicability test results
+            if (hasSingleSourceApplicableTest) {
+                if (afterWidened.size() == after.size()) {
+                    // 'afterWidened' and 'after' are 1-to-1 mapping, reflects single source applicability test results.
+                    //noinspection unchecked
+                    afterWidened = ListUtils.map((List<SourceFile>) after, (i, s) ->
+                            lastSingleSourceApplicableTestResult.get(i) ? afterWidenedRef.get(i) : s);
+                } else if (afterWidened.size() > after.size()) {
+                    newSingleSourceApplicableTestResult.addAll(Collections.nCopies(afterWidened.size() - after.size(), true));
+                }
+            }
```

Comment:
As we have been discussing elsewhere, this does not look quite right since recipes can both add and remove source files. Need something like a `Map<UUID, Boolean>` to keep track of which source files have passed which applicability tests

Reply:
Thanks for the good point! Updated the code.